### ISSUE TITLE
Allow all torch.nn modules to accept arbitrary batch dimensions

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#47149]
## 🚀 Feature

Allow all [`torch.nn`](https://pytorch.org/docs/stable/nn.html) modules (e.g. [`Conv2d`](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html), [`GRUCell`](https://pytorch.org/docs/stable/generated/torch.nn.GRUCell.html), etc.) to accept arbitrary batch dimensions, like [`Linear`](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) does.

## Motivation

This prevents having to repeatedly reshape inputs and outputs, which can be error-prone. For a simple example, when processing single inputs, one has to repeatedly add and subtract an extra dimension so the module can handle it as a batch:

```
output1, output2, output3 = module(input1[None], input2[None], input3[None])
output1 = output1[0]
output2 = output2[0]
output3 = output3[0]
```

**Comment by mruberry** [comment]
Thank you for the suggestion, @carlosgmartin.

**Comment by zou3519** [comment]
I talked about this with @cpuhrsch before -- this would make it easier for vmap and nestedtensor (separately) to work on modules if the modules could take a "single input" (an input without a batch dimension).

However, we have some modules that don't allow this with their current semantics. One example is [nn.GroupNorm](https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html) -- because it allows arbitrary feature dimensions, we cannot allow it to accept arbitrary batch dimensions by default. Modules like this are in the minority and we can probably add some flag during module construction to get around this

**Comment by carlosgmartin** [comment]
@zou3519 Out of curiosity, why don't [torch.nn.GRU](https://pytorch.org/docs/stable/generated/torch.nn.GRU.html) and [torch.nn.LSTM](https://pytorch.org/docs/stable/generated/torch.nn.LSTM.html) have the batch dimension first, like other modules do?

**Comment by zou3519** [comment]
IIRC it's more performant for the time dimension to be in the front, followed by the batch dimension

**Comment by jbschlosser** [comment]
To support vmap, I'm specifically looking into the 0-batch dim case across modules that can support this. Detailed breakdown of these modules TBD.

**Comment by jbschlosser** [comment]
Update: Check out the linked issue above for details on the torch.nn module updates required to support the no-batch-dim case.
